test-case/check-alsa-conformance.sh

```diff
@@ -0,0 +1,357 @@
+OPT_HAS_ARG['v']=0              OPT_VAL['v']=0
+
+OPT_NAME['E']='rate-diff'       OPT_DESC['E']="ALSA conformance --rate-criteria-diff-pct (difference, %)."
+OPT_HAS_ARG['E']=1              OPT_VAL['E']=''
+
+OPT_NAME['e']='rate-err'        OPT_DESC['e']="ALSA conformance --rate-err-criteria (max rate error)."
+OPT_HAS_ARG['e']=1              OPT_VAL['e']=''
+
+OPT_NAME['a']='avail-delay'     OPT_DESC['a']="ALSA conformance --avail-delay"
+OPT_HAS_ARG['a']=0              OPT_VAL['a']=0
+
+OPT_NAME['T']='test-suites'     OPT_DESC['T']="ALSA conformance --test-suites (Default: all)."
+OPT_HAS_ARG['T']=1              OPT_VAL['T']=''
+
+func_opt_parse_option "$@"
+
+# Options for the ALSA conformance test script call
+CMD_OPTS=()
+
+# Recompose OPT_VAL[$1] option as ALSA test script option $2
+add_cmd_option()
+{
+  local opt_val="${OPT_VAL[$1]}"
+  local prefix=$2
+
+  if [ -n "${opt_val}" ]; then
+    # Split list parameters to separate values
+    opt_val=("${opt_val//[ ,]/ }")
+    # shellcheck disable=SC2206
+    CMD_OPTS+=("${prefix}" ${opt_val[@]})
+  fi
+}
+
+init_globals()
+{
+  add_cmd_option 'r' '--allow-rates'
+  add_cmd_option 'F' '--allow-formats'
+  add_cmd_option 'E' '--rate-criteria-diff-pct'
+  add_cmd_option 'e' '--rate-err-criteria'
+  add_cmd_option 'T' '--test-suites'
+
+  run_verbose=0
+  if [[ "${OPT_VAL['v']}" -eq 1 ]]; then
+    run_verbose=1
+    CMD_OPTS+=("--log-file" "/dev/stdout")
+  fi
+
+  if [[ "${OPT_VAL['a']}" -eq 1 ]]; then
+    CMD_OPTS+=('--avail-delay')
+  fi
+
+  AUDIOTEST_OUT="${LOG_ROOT}/alsa_conformance"
+  RESULT_JSON="${LOG_ROOT}/test_result.json"
+
+  ALSA_CONFORMANCE_PATH=$([ -n "$ALSA_CONFORMANCE_PATH" ] || realpath "${TESTDIR}/../audiotest")
+  ALSA_CONFORMANCE_TEST="${ALSA_CONFORMANCE_PATH}/alsa_conformance_test"
+}
+
+check_alsa_conformance_suite()
+{
+  if [ -d "${ALSA_CONFORMANCE_PATH}" ]; then
+    if [ -x "${ALSA_CONFORMANCE_TEST}" ] && [ -x "${ALSA_CONFORMANCE_TEST}.py" ]; then
+      dlogi "Use ALSA conformance test suite: ${ALSA_CONFORMANCE_TEST}"
+      return
+    fi
+  fi
+  skip_test "ALSA conformance test suite is missing at: ${ALSA_CONFORMANCE_PATH}"
+}
+
+get_card_devices()
+{
+  local mode=$1 
+  local arg_pcm=$2
+
+  # select all devices by default
+  [ -z "${arg_pcm}" ] && arg_pcm="[^ ]+"
+
+  local alsa_list=''
+  local res_devs=("${arg_pcm}")
+
+  if [ "${mode}" == 'playback' ]; then
+    alsa_list=('aplay' '-l')
+  elif [ "${mode}" == 'capture' ]; then
+    alsa_list=('arecord' '-l')
+  else
+    return
+  fi
+
+  if [ -n "${arg_pcm}" ]; then
+    # check is only card name is given or exact device
+    if [ "${arg_pcm}" == "${arg_pcm##*,}" ]; then
+      # strip 'hw:' prefix
+      arg_pcm="${arg_pcm#*:}"
+      # shellcheck disable=SC2016
+      local gawk_script='match($0, /^card [0-9]+: ('"${arg_pcm}"') .+ device ([0-9]+): /, arr) { print "hw:" arr[1] "," arr[2] }'
+      mapfile -t res_devs < <( "${alsa_list[@]}" | gawk "${gawk_script}" )
+    fi
+    echo "${res_devs[@]}"
```

Comment:
For anything non trivial I would always prefer `printf`

https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo